### PR TITLE
feat(optimization): Perform array gets as early as possible

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/5027

(https://github.com/noir-lang/noir/issues/5501 is a separate issue)

## Summary\*

Due to other optimizations out of the user's control, programs can often end up with sequences resembling the following:
```
    v90208 = array_set v90198, index v90207, value u8 0
    v90210 = add v90203, u64 1
    v90211 = mul v90098, v90204
    v90212 = array_get v90208, index v90207
    v90213 = array_get v90198, index v90207
```
Where we have an `array_set` that cannot be made mutable because there is still an `array_get` later to the unmutated array.

To fix this, I've added another pass to move array gets as early as possible in the program. The earliest they can be moved is always just after their last ValueId dependency was inserted. This usually frees up any array_sets to be made mutable afterward.

## Additional Context

The program
```
fn main(foo: [u8; 95], toggle: bool) {
    let size: Field = 93 + toggle as Field * 2;
    let hash = sha256::sha256_var(foo, size as u64);
    println(f"{hash}");
}   
```
as well as the new test in `execution_success/array_get_optimization` now have 0 non-mutable array sets.

Not much change from most tests. I think this pattern was mostly a result of the array merge optimization to only merge changed indices since this is the only optimization that inserts array sets.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/opt/move_array_gets.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+//!   - The array itself, the index to set, and the enable_side_effects variable.
+//!
+//! This optimization has two passes:
+//! 1. Traverse the function to find each array_get and remember its dependencies,
+//!    draining each instruction as we go.
+//! 2. Traverse the function re-inserting each instruction. When all dependencies
+//!    of an array_get have been inserted we can insert the corresponding array_get itself.
+//!    Since it can be expensive checking this for every instruction, this pass makes
+//!    the assumption that the last dependency will always have the highest ValueId.
+use fxhash::FxHashMap;
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/opt/move_array_gets.rs:21]:
```suggestion
use fxhash::FxHashMap as HashMap;
```
nit: just for consistency with what we do throughout the codebase

**Review by vezenovm — APPROVED** [review]

File: compiler/noirc_evaluator/src/ssa/opt/move_array_gets.rs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                move_array_gets(state, &mut func.dfg, block, instructions);
+            }
+        }
+        self
+    }
+}
+
+#[derive(Default)]
+struct State {
+    array_gets: FxHashMap<ValueId, Vec<ArrayGet>>,
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/opt/move_array_gets.rs:48]:
```suggestion
    array_gets: HashMap<ValueId, Vec<ArrayGet>>,
```

Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/opt/move_array_gets.rs:48]: And could we also add a comment here just explaining this map and that the key is the last dependency?

Comment (jfecher) [reply compiler/noirc_evaluator/src/ssa/opt/move_array_gets.rs]: Done

**Review by TomAFrench — CHANGES_REQUESTED** [review]
Requesting changes just for some discussion. 

I'm a little concerned about this PR as it's similar to my "bubble up" optimisation which does the same operation but for constrain instructions. This resulted in situations where we ended up getting tests failing due to the noir code not being executed in the order which a user would expect.

I'm unsure how long we can maintain this guarantee however so we may need to open discussions on breaking this.

**Comment by jfecher** [comment]
@TomAFrench I'm going to try changing the array merge optimization directly for now to delay the array sets there instead of moving the array gets later. This is the first approach I tried and thought it didn't work initially but I think now that the noir program was just getting cached from a previous version of the compiler. So I'll try again and hopefully we'll see similar improvements.

I'll also note that the conditional_1 test failure did not occur when I used a smaller constant value for that array merge optimization, like the previous 2. Not sure why that'd be causing the failure.

**Comment by jfecher** [comment]
Running into some difficulties experimenting with this PR. Most notably is a difference in semantics between SSA Pre and Post-Flattening.

When optimizing array sets we optimize them if the array and index are known, but we never check the current enable side effects value to see if the array set may be disabled. This is fine Pre-flattening since control flow prevents referring to these directly if it is not what is wanted. Post flattening however it should be valid to optimize an array merge of the following:
```
b0(v1: [Field; 2], v2: u1):
  enable_side_effects v2
  v3 = array_set v1, index 0, value 1
  v4 = not v2
  enable_side_effects v4
  v5 = array_set v1, index 1, value 2
  enable_side_effects u1 1
  v6 = if v2 then v3 else if v4 then v5
```
to:
```
b0(v1: [Field; 2], v2: u1):
  enable_side_effects v2
  v3 = array_set v1, index 0, value 1
  v4 = not v2
  enable_side_effects v4
  v5 = array_set v3, index 1, value 2  // changed v1 to v3
  enable_side_effects u1 1
  // v6 = v5
```
Since the then branch should be disabled in the else case, we should expect `v3 = v1`, and can set to that directly instead. Likewise, since the else branch is disabled in the then case, we don't have to worry about `v5` changing the result of the merge in that case. This also lets us write mutably in both branches. But I believe this doesn't currently work when `v1` is constant because the else branch will be optimized to a constant `[_, 2]` even if it was not meant to be executed.

**Comment by jfecher** [comment]
Going to pause work on this PR until https://github.com/noir-lang/noir/issues/5782 is fixed. I'm also running into it with the current approach in this PR which is fairly similar to what we currently use. The nested dynamic arrays test is failing with this panic currently.

**Comment by jfecher** [comment]
Hmm I'm surprised changing the array merge optimization to use `[array1, array2][condition as u32]` is so much less efficient than merging every index of the arrays and creating a new array.

**Comment by jfecher** [comment]
I'm closing this for now since the results are currently negative. We can retry similar optimizations in another PR later if wanted.

**Pull request closed without merging**